url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/triggers/CronExpression.java

```diff
@@ -0,0 +1,1571 @@
+        i++;
+        if (i >= s.length()) {
+          addToSet(val, end, v2, type);
+          return i;
+        }
+        c = s.charAt(i);
+        if (c >= '0' && c <= '9') {
+          ValueSet vs = getValue(v2, s, i);
+          int v3 = vs.value;
+          addToSet(val, end, v3, type);
+          i = vs.pos;
+          return i;
+        } else {
+          addToSet(val, end, v2, type);
+          return i;
+        }
+      } else {
+        addToSet(val, end, 1, type);
+        return i;
+      }
+    }
+
+    if (c == '/') {
+      i++;
+      c = s.charAt(i);
+      int v2 = Integer.parseInt(String.valueOf(c));
+      i++;
+      if (i >= s.length()) {
+        addToSet(val, end, v2, type);
+        return i;
+      }
+      c = s.charAt(i);
+      if (c >= '0' && c <= '9') {
+        ValueSet vs = getValue(v2, s, i);
+        int v3 = vs.value;
+        addToSet(val, end, v3, type);
+        i = vs.pos;
+        return i;
+      } else {
+        throw new ParseException("Unexpected character '" + c + "' after '/'", i);
+      }
+    }
+
+    addToSet(val, end, 0, type);
+    i++;
+    return i;
+  }
+
+  public String getCronExpression() {
+    return cronExpression;
+  }
+
+  public String getExpressionSummary() {
+    StringBuilder buf = new StringBuilder();
+
+    buf.append("seconds: ");
+    buf.append(getExpressionSetSummary(seconds));
+    buf.append("\n");
+    buf.append("minutes: ");
+    buf.append(getExpressionSetSummary(minutes));
+    buf.append("\n");
+    buf.append("hours: ");
+    buf.append(getExpressionSetSummary(hours));
+    buf.append("\n");
+    buf.append("daysOfMonth: ");
+    buf.append(getExpressionSetSummary(daysOfMonth));
+    buf.append("\n");
+    buf.append("months: ");
+    buf.append(getExpressionSetSummary(months));
+    buf.append("\n");
+    buf.append("daysOfWeek: ");
+    buf.append(getExpressionSetSummary(daysOfWeek));
+    buf.append("\n");
+    buf.append("lastdayOfWeek: ");
+    buf.append(lastdayOfWeek);
+    buf.append("\n");
+    buf.append("nearestWeekday: ");
+    buf.append(nearestWeekday);
+    buf.append("\n");
+    buf.append("NthDayOfWeek: ");
+    buf.append(nthdayOfWeek);
+    buf.append("\n");
+    buf.append("lastdayOfMonth: ");
+    buf.append(lastdayOfMonth);
+    buf.append("\n");
+    buf.append("years: ");
+    buf.append(getExpressionSetSummary(years));
+    buf.append("\n");
+
+    return buf.toString();
+  }
+
+  protected String getExpressionSetSummary(java.util.Set<Integer> set) {
+
+    if (set.contains(NO_SPEC)) {
+      return "?";
+    }
+    if (set.contains(ALL_SPEC)) {
+      return "*";
+    }
+
+    StringBuilder buf = new StringBuilder();
+
+    Iterator<Integer> itr = set.iterator();
+    boolean first = true;
+    while (itr.hasNext()) {
+      Integer iVal = itr.next();
+      String val = iVal.toString();
+      if (!first) {
+        buf.append(",");
+      }
+      buf.append(val);
+      first = false;
+    }
+
+    return buf.toString();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 816:837, 842:863.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.